external/llvm-project/mlir/lib/Conversion/XModelToGPU/XModelToGPU.cpp

```diff
@@ -91,45 +91,79 @@ struct LaunchRewritePattern : public OpRewritePattern<async::LaunchOp> {
     return b.create<gpu::WaitOp>(loc, tokenType, deps).asyncToken();
   }
 
-  Value moveMemory(OpBuilder b, Value opr, uint32_t fidx, bool readAccess,
-                   bool writeAccess, llvm::SmallVector<Value> &copyBackOprs,
+  template <typename T>
+  bool isOnDevice(const T &oprUsers) const {
+    for (auto opUse : oprUsers) {
+      auto gpuLaunch = dyn_cast<gpu::LaunchFuncOp>(opUse);
+      auto launch = dyn_cast<async::LaunchOp>(opUse);
+      // assumes the same GPU
+      if (!gpuLaunch && !(launch && getGPUTarget(launch).has_value()))
+        return false;
+    }
+    return true;
+  }
+
+  Value moveMemory(OpBuilder b, async::LaunchOp launchOp, Value opr,
+                   uint32_t fidx, bool readAccess, bool writeAccess,
+                   llvm::SmallVector<Value> &copyBackOprs,
                    llvm::SmallVector<Value, 8> &asyncDeps) const {
+    if (auto gpuAllocOp = opr.getDefiningOp<gpu::AllocOp>()) {
+      // TEST: convergence or multi-input??
+      assert(isOnDevice(opr.getUsers()));
```

Comment:
It should always be true, I think. I think the tests now cover all cases.

Reply:
Ok, cool, I just want to avoid crashing the program on things that you're likely to encounter in well-formed input someone might reasonably expect to work.